feat: jwks caching with ActiveSupport

vanessa-passage opened the pull request:
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

## What's New?

1. Removes the refetch of jwks when call fails.
2. Implements caching with ActiveSupport

## Screenshots (if appropriate):

<!-- Please attach any screenshots that would help illustrate the changes. -->

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have manually tested my code thoroughly
- [ ] I have added/updated inline documentation for public facing interfaces if relevant
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing integration and unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

## Additional context

<!-- Add any other context or screenshots about the pull request here. -->

ctran88 commented:
what do you think of using something like `ruby-jwt` (which we already have as a dependency) to manage the [jwks caching](https://github.com/jwt/ruby-jwt#json-web-key-jwk)? here's a [blog](https://unathi.hashnode.dev/verifying-jwts-with-jwks-in-ruby) describing the implementation as well.

vanessa-passage commented:
> what do you think of using something like `ruby-jwt` (which we already have as a dependency) to manage the [jwks caching](https://github.com/jwt/ruby-jwt#json-web-key-jwk)? here's a [blog](https://unathi.hashnode.dev/verifying-jwts-with-jwks-in-ruby) describing the implementation as well.

From my understanding, they are using Rails built in cache. 
```
    def jwks(force: false)
      Rails.cache.fetch(JWKS_CACHE_KEY, force: force, skip_nil: true) do
        fetch_jwks
      end&.deep_symbolize_keys
    end
```

We don't have access to this method of caching because we aren't using Rails. I have introduced the `active_support` gem to provide the ability to cache to Ruby. This was what I mentioned in check in today that was driving me crazy. I kept trying to access Rails solutions 🤦‍♀️ . With that being said I still need to make some changes to how I implemented Active Support.

ctran88 commented:
> they are using Rails built in cache

Sorry I should have clarified; with the blog link I just meant to reference how they're fetching the remote JWKS (ignoring the rails implementation since we're not using it).

For the caching implementation, does [this part](https://github.com/jwt/ruby-jwt#json-web-key-jwk:~:text=jwks_loader%20%3D%20%2D%3E,jwks%0A%20%20end%0Aend) of the ruby-jwt example not work for us?

vanessa-passage commented:
> > they are using Rails built in cache
> 
> Sorry I should have clarified; with the blog link I just meant to reference how they're fetching the remote JWKS (ignoring the rails implementation since we're not using it).
> 
> For the caching implementation, does [this part](https://github.com/jwt/ruby-jwt#json-web-key-jwk:~:text=jwks_loader%20%3D%20%2D%3E,jwks%0A%20%20end%0Aend) of the ruby-jwt example not work for us?

I will look into it! There are a few improvements that can be added to this. I will look into the link you provided, I hadn't seen this option.

vanessa-passage commented on Gemfile:
```diff
@@ -6,6 +6,7 @@ source 'https://rubygems.org'
 gemspec
 
 group :development do
+  gem 'activesupport', '~> 7.2'
```
The most recent version of this gem is 8.0. Ruby version needs to be increased to 3.2.

ctran88 replied:
I think this should be moved to the [gemspec](https://github.com/passageidentity/passage-ruby/blob/cad1b0591537b93cab13babebab6c21d2179cd09/passageidentity.gemspec#L26) instead since this section is only for dev dependencies

vanessa-passage replied:
100% good catch.

vanessa-passage commented on lib/passageidentity/auth.rb:
```diff
@@ -14,7 +15,7 @@ class Auth
     # rubocop:disable Metrics/AbcSize
     def initialize(app_id, api_key, auth_strategy)
-      @app_cache = {}
+      @app_cache = ActiveSupport::Cache::MemoryStore.new
```
@ctran88 It is still necessary to use an additional caching library.

ctran88 commented on lib/passageidentity/auth.rb:
```diff
@@ -221,11 +234,11 @@ def user_exists?(user_id)
-    def set_cache(key, value)
-      @app_cache[key] = value
+    def set_cache(key:, jwks:)
+      @app_cache.write(key, jwks, expires_in: 3600)
```
This can be a little more lax. Our node SDK expires at [24 hours](https://github.com/passageidentity/passage-node/blob/c94769c59e28139d7546cd0335dc4bf4476ccf50/src/classes/Auth/Auth.ts#L29)

ctran88 commented on lib/passageidentity/auth.rb:
```diff
@@ -74,6 +75,16 @@ def validate_jwt(token)
       exists = jwk_exists(token)
```
Looks like the refetch is still happening if the kid is missing
